File: backend-app/src/main/java/com/project/gamereviewer/entity/Game.java
package com.project.gamereviewer.entity;

import java.util.HashSet;
import java.util.Set;

import org.hibernate.annotations.Formula;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

@Entity
@Table(name = "games", uniqueConstraints = {
    @UniqueConstraint(columnNames = {"title", "release_year"})
})
@Getter
@Setter
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
@NoArgsConstructor
@AllArgsConstructor
@ToString(exclude = {"genres", "reviews", "systemRequirements"})
public class Game {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @EqualsAndHashCode.Include
    private Integer id;

    @Column(nullable = false, length = 150)
    private String title;

    @Column(name = "release_year")
    private Integer releaseYear;

    @Column(columnDefinition = "TEXT")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "developer_id")
    private ProductionCompany developer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "publisher_id")
    private ProductionCompany publisher;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
        name = "games_genres",
        joinColumns = @JoinColumn(name = "game_id"),
        inverseJoinColumns = @JoinColumn(name = "genre_id")
    )
    private Set<Genre> genres = new HashSet<>();

    @OneToMany(mappedBy = "game", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private Set<Review> reviews = new HashSet<>();

    @OneToMany(mappedBy = "game", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private Set<SystemRequirement> systemRequirements = new HashSet<>();
    
    @Formula("(SELECT COALESCE(AVG(r.score), 0.0) FROM reviews r WHERE r.game_id = id)")
    private Double averageRating;
}

File: frontend-angular-app/src/app/core/models/game.ts
import { Genre } from './genre';
import { ProductionCompany } from './production-company';
import { Review } from './review';
import { SystemRequirement } from './system-requirement';

export interface GameListItem {
  id: number;
  title: string;
  releaseYear: number | null;
  developerName: string | null;
  publisherName: string | null;
  genreNames: string[];
  averageRating: number | null;
}

export interface GameDetail {
  id: number;
  title: string;
  releaseYear: number | null;
  description: string | null;
  developer: ProductionCompany | null;
  publisher: ProductionCompany | null;
  genres: Genre[];
  systemRequirements: SystemRequirement[];
  reviews: Review[];
  averageRating: number | null;
}

export interface GameCreateRequest {
  title: string;
  releaseYear: number;
  description?: string | null;
  developerId?: number | null;
  publisherId?: number | null;
  genreIds: number[];
}

export interface GameUpdateRequest {
  title?: string | null;
  releaseYear?: number | null;
  description?: string | null;
  developerId?: number | null;
  publisherId?: number | null;
  genreIds?: number[] | null;
}
